### PR TITLE
🐛 default service monitor configuration to use https

### PR DESCRIPTION
**Pull request by johanneswuerbach**
By default the ServiceMonitor scrapes `port: https` https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v2/config/prometheus/monitor.yaml#L13, which is exposed by the auth proxy service https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v2/config/rbac/auth_proxy_service.yaml#L10-L12 and requires an https connection https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v2/config/default/manager_auth_proxy_patch.yaml#L15

Change the ServiceMonitor configuration to work with that by default as otherwise the scraping fails with a `400 Bad Request`.

Thanks to https://github.com/kubernetes-sigs/kubebuilder/issues/1253#issuecomment-597640633 for the required flags.

### CODE REVIEW

File: pkg/plugins/golang/v2/scaffolds/internal/templates/config/prometheus/monitor.go
Comment (estroz) [inline pkg/plugins/golang/v2/scaffolds/internal/templates/config/prometheus/monitor.go:59]: At this point, no new features should go into go/v2
```suggestion
```

File: pkg/plugins/golang/v3/scaffolds/internal/templates/config/prometheus/monitor.go
Comment (estroz) [inline pkg/plugins/golang/v3/scaffolds/internal/templates/config/prometheus/monitor.go:59]: The webhook server already [sets up TLS](https://github.com/kubernetes-sigs/controller-runtime/blob/774f9d490254c861e9f5ca009b3c4b8b5d55fa83/pkg/webhook/server.go#L175). I wonder if that cert/key can be used here.

File: pkg/plugins/golang/v2/scaffolds/internal/templates/config/prometheus/monitor.go
Comment (johanneswuerbach) [reply pkg/plugins/golang/v2/scaffolds/internal/templates/config/prometheus/monitor.go:59]: As the current configuration doesn’t work, it could also be seen as a bug fix, but happy to remove this part.

Comment (estroz) [reply pkg/plugins/golang/v2/scaffolds/internal/templates/config/prometheus/monitor.go:59]: If that's true, then it can be left in I suppose.

Comment (Adirio) [reply pkg/plugins/golang/v2/scaffolds/internal/templates/config/prometheus/monitor.go:59]: If it is a bugfix (haven't checked that) it should have the 🐛 (`:bug:`) emoji.

File: docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml
Comment (camilamacedo86) [inline docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml:18]: I do not think that we should move forward here. See the doc: https://book.kubebuilder.io/reference/metrics.html

These metrics are protected by kube-auth-proxy by default if using kubebuilder. Kubebuilder v2.2.0+ scaffold a clusterrole which can be found at config/rbac/auth_proxy_client_clusterrole.yaml.

```
You will need to grant permissions to your Prometheus server so that it can scrape the protected metrics. To achieve that, you can create a clusterRoleBinding to bind the clusterRole to the service account that your Prometheus server uses.

You can run the following kubectl command to create it. If using kubebuilder <project-prefix> is the namePrefix field in config/default/kustomization.yaml.

kubectl create clusterrolebinding metrics --clusterrole=<project-prefix>-metrics-reader --s
```

Did you follow that?

Comment (camilamacedo86) [reply docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml:18]: /hold

Comment (johanneswuerbach) [reply docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml:18]: Yes, but I'm not sure how your comment is related.

The proxy is [by default](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v2/config/default/manager_auth_proxy_patch.yaml#L15) exposed only via `https` so scraping using `scheme: http` [will result in a 400 error](https://github.com/golang/go/blob/520f3b72db7befab2028d9a47376267cf2d274a9/src/net/http/server.go#L1844-L1845). Annoyingly this directly implemented in golang, so the rbac proxy doesn't log anything and prometheus doesn't log the response body.

The `bearerTokenFile` is required for prometheus to actually [include a](https://github.com/prometheus-operator/prometheus-operator/blob/da324dd3d75ce7db24c6c55704a2231d1249fec6/pkg/prometheus/promcfg.go#L916) `Authorization` [header](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config).

This header is required so the proxy can actually validate whether the request has the necessary permissions (you get a [401 otherwise, while the missing binding would result in a 403](https://github.com/brancz/kube-rbac-proxy/blob/e4b31758aedb3d29d8f50d615828401a30d28c1e/pkg/proxy/proxy.go#L62))

The `insecureSkipVerify` is required as the rbac proxy uses a self-generate certificate in the current configuration https://github.com/brancz/kube-rbac-proxy/blob/e4b31758aedb3d29d8f50d615828401a30d28c1e/main.go#L261

@estroz proposed to use the webhook certificate https://github.com/kubernetes-sigs/kubebuilder/pull/2065#discussion_r588711493, but this seems out of scope of this PR.

File: pkg/plugins/golang/v3/scaffolds/internal/templates/config/prometheus/monitor.go
Comment (johanneswuerbach) [reply pkg/plugins/golang/v3/scaffolds/internal/templates/config/prometheus/monitor.go:59]: This seems out of scope of this PR, but seems like a good idea otherwise.

File: pkg/plugins/golang/v2/scaffolds/internal/templates/config/prometheus/monitor.go
Comment (johanneswuerbach) [reply pkg/plugins/golang/v2/scaffolds/internal/templates/config/prometheus/monitor.go:59]: I wasn't sure, but as the default configuration just doesn't work I think it makes sense.

File: docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml
Comment (camilamacedo86) [reply docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml:18]: I can be missing something here our misunderstanding. However, see: https://github.com/kubernetes-sigs/kubebuilder/pull/1317/files#diff-20a08dcfd7ad4ed3343d00cf32c7844ce41a2d8d5f5df5b6de56852a510c5517R46 from the PR: https://github.com/kubernetes-sigs/kubebuilder/pull/1317 which via the rbac rules allows accessing metrics behind kube-rbac-proxy. 

Then, see that in the e2e tests that are checked: 

- https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/v3/plugin_cluster_test.go#L165-L170
- https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/v3/plugin_cluster_test.go#L272-L303

So, I am trying here understand why in the e2e test we get HTTP 200 from the metrics endpoint and you are facing 400 or 403. What is missing? What has been working our e2e tests and not for you? 

Is it because we are getting the [token](https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/v3/plugin_cluster_test.go#L265-L270) to check and doing the handshake and you are not?  

Then, I understand that the default configuration has been allowing Prometheus to get the metrics. See that the problem point out by you was solved in the past with the rbac roles https://github.com/kubernetes-sigs/kubebuilder/pull/1317.

Anyway, shows that also has some difference between using Prometheus Operator instead of kube-prometheus which has been solved with the solution applied here as you point out : https://github.com/kubernetes-sigs/kubebuilder/issues/1253#issuecomment-597640633 

So, I will 

/hold cancel 

However, would be very nice to have a new issue with the steps required to reproduce the scenario for the future we are able to check that. Describing the cluster and Prometheus used, and if the rbac roles for allowing the Prometheus get the metrics were or not used, and what are the steps to check the error.

File: testdata/project-v3/config/prometheus/monitor.yaml
Comment (camilamacedo86) [inline testdata/project-v3/config/prometheus/monitor.yaml:17]: Would not it remove the need to get the token in the e2e tests? 

- https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/v3/plugin_cluster_test.go#L165-L170
- https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/v3/plugin_cluster_test.go#L272-L303

If yes, I think we need to remove that from the e2e tests to validate your solution as well.

File: docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml
Comment (Adirio) [reply docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml:18]: Actually, Camila's point is pretty sound. If e2e tests are working and this is being tested, we need an issue describing the steps to get this error so that we can reproduce and check that it is an actual bug and not just an error on your side.

Comment (johanneswuerbach) [reply docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml:18]: Looking at the e2e tests, it seems they only test the service endpoint using curl, but the service monitor itself is not covered. The service monitor instructs how prometheus is supposed to scrape the service endpoint.

The test itself is using https (which is configured via scheme), `-k` which skips the certificate validation and injects a token into the request, which seems to be confirming the service monitor misconfiguration.

I‘m happy to put together a reproduction, but it requires prometheus so might be not something usable for e2e tests.

File: testdata/project-v3/config/prometheus/monitor.yaml
Comment (johanneswuerbach) [reply testdata/project-v3/config/prometheus/monitor.yaml:17]: No, this only instructs how prometheus should scrape the service, but doesn’t change anything about the service itself. As prometheus isn’t used in e2e tests this doesn’t really change anything as this part is not test covered yet.

File: docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml
Comment (Adirio) [reply docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml:18]: > The test itself is using https (which is configured via scheme), `-k` which skips the certificate validation and injects a token into the request, which seems to be confirming the service monitor misconfiguration.

That actually makes sense. I verified it and we are exactly testing it as you describe.

> I‘m happy to put together a reproduction, but it requires prometheus so might be not something usable for e2e tests.

The reproduction we were talking about, at least from my part, was how to reproduce the error you fix here. But changing the `curl` approach for a real e2e test would be pretty good to avoid making the same mistake in the future. Why do you think that requiring prometheus is not suitable for e2e?

Comment (camilamacedo86) [reply docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml:18]: > I‘m happy to put together a reproduction, but it requires prometheus so might be not something usable for e2e tests.

We install Prometheus in the e2e tests see: https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/v3/plugin_cluster_test.go#L45-L46 

So, I think we need also cover this scenario to be sure that indeed this solution will fix it and we will not break that in the future. Could you help with that?

Comment (johanneswuerbach) [reply docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml:18]: I'm happy to see what can be done, currently I'm struggling with getting the e2e test suite running on the new M1 MacBook, so an update here might require a bit more time.

Comment (Adirio) [reply docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml:18]: Go 1.16? Not supported yet. There are some modules related changes (like some commands not tidying up go.mod anymore) that haven't been taking into account in kubebuilder nor any of their dependencies.

Comment (estroz) [reply docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml:18]: Ah that is unfortunate. If I have some time I can update e2e code after this is merged.

Comment (camilamacedo86) [reply docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml:18]: Just to highlight: We might be unable to reproduce the issue in the e2e as it now if it is specific only when to use Prometheus Operator instead of kube-prometheus.

**Review by camilamacedo86 — APPROVED** [review]
It shows OK since more than one person shows that have this problem and solves it the same way.  We might be unable to reproduce the issue in the e2e as it now if it is specific only when to use Prometheus Operator instead of kube-prometheus.

So, it is fine for me. 

More info: https://github.com/kubernetes-sigs/kubebuilder/pull/2065/files#r589002353

File: docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml
Comment (Adirio) [reply docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml:18]: If you create an issue describing the need of an e2e test that covers this case, I'm ok merging this.

Comment (johanneswuerbach) [reply docs/book/src/multiversion-tutorial/testdata/project/config/prometheus/monitor.yaml:18]: Opened https://github.com/kubernetes-sigs/kubebuilder/issues/2087

**Comment by johanneswuerbach** [comment]
I created the related e2e test ticket https://github.com/kubernetes-sigs/kubebuilder/issues/2087, anything else missing here?

**Review by estroz — APPROVED** [review]
Since the metrics endpoint is protected by kube-rbac-proxy, I don't see a reason against merging this.

/lgtm

**Pull request merged**
